src/sentry/api/serializers/models/release.py

```diff
@@ -311,16 +322,44 @@ def get_attrs(self, item_list, user, **kwargs):
         if with_health_data and no_snuba:
             raise TypeError("health data requires snuba")
 
+        self.has_adoption_stage = features.has(
+            "organizations:release-adoption-stage", item_list[0].organization_id
+        )
+        if self.has_adoption_stage:
+            release_project_envs = (
+                ReleaseProjectEnvironment.objects.filter(release__in=item_list)
+                .select_related("release")
+                .order_by("-first_seen")
+            )
+            if environments is not None:
+                release_project_envs = release_project_envs.filter(
+                    environment__name__in=environments
+                )
+            if project is not None:
+                release_project_envs = release_project_envs.filter(project=project)
+
+            adoption_stages = self._get_release_adoption_stages(release_project_envs)
+
         if environments is None:
             first_seen, last_seen, issue_counts_by_release = self.__get_release_data_no_environment(
                 project, item_list
             )
         else:
+            if not self.has_adoption_stage:
+                release_project_envs = (
+                    ReleaseProjectEnvironment.objects.filter(
+                        release__in=item_list, environment__name__in=environments
+                    )
+                    .select_related("release")
+                    .order_by("-first_seen")
+                )
+                if project is not None:
+                    release_project_envs = release_project_envs.filter(project=project)
```

Comment:
This felt a bit awkward, but since I moved the `ReleaseProjectEnvironment` query into this function to use it in both `_get_release_adoption_stages` and `__get_release_data_with_environments`, it would run for every customer that didn't have the feature. So I just added this branching logic for now to avoid that